flink-external-resource/flink-external-resource-gpu/src/main/java/org/apache/flink/externalresource/gpu/GPUDriver.java

```diff
@@ -0,0 +1,134 @@
+	static final ConfigOption<String> DISCOVERY_SCRIPT_ARG =
+		key(ExternalResourceOptions.keyWithResourceNameAndSuffix(RESOURCE_NAME, DISCOVERY_SCRIPT_ARGS_SUFFIX))
+			.stringType()
+			.noDefaultValue();
+
+	@VisibleForTesting
+	static final ConfigOption<Long> GPU_AMOUNT =
+		key(ExternalResourceOptions.keyWithResourceNameAndSuffix(RESOURCE_NAME, ExternalResourceOptions.EXTERNAL_RESOURCE_AMOUNT_SUFFIX))
+			.longType()
+			.defaultValue(0L);
+
+	private final Set<GPUInformation> gpuResources;
+
+	public GPUDriver(Configuration config) throws Exception {
+		final String discoveryScriptPath = config.getString(DISCOVERY_SCRIPT_PATH);
+		if (StringUtils.isNullOrWhitespaceOnly(discoveryScriptPath)) {
+			throw new FlinkRuntimeException("Could not find config of the path of gpu discovery script.");
+		}
+
+		final File discoveryScript = new File(System.getenv().getOrDefault(ConfigConstants.ENV_FLINK_HOME_DIR, ".") +
+			"/" + discoveryScriptPath);
+		if (!discoveryScript.exists()) {
+			throw new FlinkRuntimeException("The gpu discovery script does not exist in path " + discoveryScript.getAbsolutePath());
+		}
+
+		final String args = config.getString(DISCOVERY_SCRIPT_ARG);
+		final long gpuAmount = config.getLong(GPU_AMOUNT);
+		gpuResources = new HashSet<>();
+
+		if (gpuAmount <= 0) {
+			LOG.warn("The amount of GPU should be positive.");
+			return;
+		}
+
+		String output = executeDiscoveryScript(discoveryScript, gpuAmount, args);
+		if (output != null && output != "") {
+			String[] indexes = output.split(",");
+			for (String index : indexes) {
+				gpuResources.add(new GPUInformation(index));
+			}
+		}
+		LOG.info("Discover the GPU resource {}.", gpuResources);
+	}
+
+	@Override
+	public Set<GPUInformation> retrieveResourceInfo(long gpuAmount) {
+		return Collections.unmodifiableSet(gpuResources);
+	}
+
+	private String executeDiscoveryScript(File discoveryScript, long gpuAmount, String args) throws Exception {
+
+		final StringBuilder cmdBuilder = new StringBuilder();
+		cmdBuilder
+			.append(discoveryScript.getAbsolutePath())
+			.append(" ")
+			.append(gpuAmount)
+			.append(" ")
+			.append(args);
+
+		final Process process = Runtime.getRuntime().exec(cmdBuilder.toString());
+		final BufferedReader stdoutReader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+		final BufferedReader stderrReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+		final int exitVal = process.waitFor();
+		if (exitVal != 0) {
+			final String stdout = stdoutReader.lines().collect(StringBuilder::new, StringBuilder::append, StringBuilder::append).toString();
+			final String stderr = stderrReader.lines().collect(StringBuilder::new, StringBuilder::append, StringBuilder::append).toString();
+			LOG.error("Discovery script exit with {}. With output {} {}", exitVal, stdout, stderr);
+			throw new FlinkRuntimeException("Discovery script exit with non-zero.");
+		}
+		return stdoutReader.readLine();
```

Comment:
I think it would be good to deduplicate the process execution logic in Flink. However, after a quick search, this logic is involved in a lot of code paths in the current Flink. So, I prefer to move this refactor out of the scope of this PR.

Reply:
This makes sense. Then let's open a follow-up ticket to track this.